### ISSUE TITLE
Groovy 2.5.5 unreachable line in byte code

### ISSUE THREAD

**Issue by admxiii** [jacoco/jacoco#884]
JaCoCo version: 0.8.4

Starting with groovy 2.5.5 through 2.5.7, an additional line number was added to the byte code to fix some issues. This causes jacoco to report lines that cannot ever be covered. This new line number seems to cause problems anytime it is after a ARETURN statement on the closing brace of the method. It would be great if these unreachable lines could be filtered out somehow.

int nonVoidMethod() {
       1 * 1
} <-- shows red

Example of new byte code 
<snip>
ARETURN
 L1
    LINENUMBER 16 L1     <-- new line number that cant be hit
   FRAME FULL [] [java/lang/Throwable]
    NOP
    ATHROW
    LOCALVARIABLE this Lcom/company/Form; L0 L1 0
    MAXSTACK = 2
    MAXLOCALS = 1

groovy 2.5.4 byte code
 <snip>
   ARETURN
   L1
   FRAME FULL [] [java/lang/Throwable]
    NOP
    ATHROW
    LOCALVARIABLE this Lcom/company/Form; L0 L1 0
    MAXSTACK = 2
    MAXLOCALS = 1

related to fixed issues
https://issues.apache.org/jira/browse/GROOVY-7647
https://issues.apache.org/jira/browse/GROOVY-8742

**Comment by Godin** [comment]
Even in Groovy 2.5.4, there is unreachable instructions, which shows up in coverage reports:

![groovy-2 5 4-class](https://user-images.githubusercontent.com/138671/58039264-89ae6600-7b32-11e9-9435-347b259ea2ad.png)

![groovy-2 5 4-source](https://user-images.githubusercontent.com/138671/58039270-8d41ed00-7b32-11e9-8dd2-094996abae45.png)

Exactly the same instructions shows up with Groovy 2.5.7 just with different line number:

![groovy-2 5 7-class](https://user-images.githubusercontent.com/138671/58039279-916e0a80-7b32-11e9-9687-f5a6e295927e.png)

![groovy-2 5 7-source](https://user-images.githubusercontent.com/138671/58039281-9337ce00-7b32-11e9-8b82-26535bddc551.png)

And this was already reported to developers of Groovy compiler some time ago - http://groovy.329449.n5.nabble.com/Groovy-2-5-4-generates-dead-code-td5755188.html
With conclusion that this should be fixed in compiler - see reply from @danielsun1106

So @admxiii please report this to developers of Groovy to remind them about this.

**Comment by admxiii** [comment]
ok thanks, I will open a ticket with Groovy

**Comment by Godin** [comment]
@admxiii thank you! 👍 ❤️

**Comment by Godin** [comment]
@admxiii I'm wondering if you managed to create ticket on Groovy side, so that we can track it too?

@aalmiray FYI if you need something more interesting than just writing tests at Hackergarten 😉

**Comment by admxiii** [comment]
Here's the ticket I created - https://issues.apache.org/jira/browse/GROOVY-9126

**Comment by danielsun1106** [comment]
@admxiii The issue will be fixed in Groovy 3.0.0-rc-2 and Groovy 2.5.9. 
Here is the PR to review and merge:  https://github.com/apache/groovy/pull/1095

**Comment by Godin** [comment]
@danielsun1106 that's cool! and thank you for the notification :thumbsup: ❤️ 

I wanted to test `2.5.9-SNAPSHOT` on JaCoCo validation tests, however could not find it in https://oss.jfrog.org/artifactory/oss-snapshot-local/org/codehaus/groovy/ , which is according to https://groovy-lang.org/install.html supposed to contain snapshot builds. Do you know why it is not there?

**Comment by danielsun1106** [comment]
Groovy core team's CI server is down, so no SNAPSHOT is available, but you could clone the project ( https://github.com/apache/groovy.git ) and switch to 2_5_X branch, run `gradlew installGroovy` to get SNAPSHOT distribution ;-)

**Comment by Godin** [comment]
> you could clone the project ( https://github.com/apache/groovy.git ) and switch to 2_5_X branch, run `gradlew installGroovy` to get SNAPSHOT distribution ;-)

@danielsun1106 I know, just wanted a bit of laziness during vacation 😆 Will do given that CI server is down. Thank you.

**Comment by danielsun1106** [comment]
Enjoy your vacation :-)

**Comment by danielsun1106** [comment]
Tested groovy code coverage with IntelliJ IDEA. As we can see, no red mark at the tailing of the method:

![image](https://user-images.githubusercontent.com/9151616/69904801-13929200-13e6-11ea-9957-dcb67ccbdbca.png)

**Comment by Godin** [comment]
@danielsun1106 I made a test of `2.5.9-SNAPSHOT` built from `GROOVY_2_5_X` branch at apache/groovy@c37ec7e2e048ddabd860967ed4b6130b4f62eec7 :

According to this test unreachable instructions do not have line numbers anymore,
however instructions are still there as was reported in http://groovy.329449.n5.nabble.com/Groovy-2-5-4-generates-dead-code-td5755188.html , which is mentioned in https://github.com/jacoco/jacoco/issues/884#issuecomment-494075741 here.

AFAIK IntelliJ computes and shows line coverage, i.e. line is considered as covered when at least one instruction belonging to this line is covered. While JaCoCo also computes instructions coverage and shows lines as partially covered when not all instructions are covered. And so while change in Groovy helps for line coverage, instructions coverage is still affected by presence of dead untestable code.

See [example.zip](https://github.com/jacoco/jacoco/files/3917571/example.zip) for which execution of `mvn verify` produces following JaCoCo report

![class report](https://user-images.githubusercontent.com/138671/70068678-bcacd880-15f0-11ea-95c8-01f820e9392f.png)

![file report](https://user-images.githubusercontent.com/138671/70068687-c0d8f600-15f0-11ea-8d1b-df394c0a71dd.png)

and here is output of `javap -v -p target/classes/Example.class` for method `non_void_method`

```
  public static int non_void_method();
    descriptor: ()I
    flags: ACC_PUBLIC, ACC_STATIC
    Code:
      stack=3, locals=1, args_size=0
         0: invokestatic  #20                 // Method $getCallSiteArray:()[Lorg/codehaus/groovy/runtime/callsite/CallSite;
         3: astore_0
         4: aload_0
         5: ldc           #33                 // int 0
         7: aaload
         8: ldc           #2                  // class Example
        10: ldc           #35                 // String Hello, World!
        12: invokeinterface #41,  3           // InterfaceMethod org/codehaus/groovy/runtime/callsite/CallSite.callStatic:(Ljava/lang/Class;Ljava/lang/Object;)Ljava/lang/Object;
        17: pop
        18: bipush        42
        20: ireturn
        21: nop
        22: nop
        23: athrow
      StackMapTable: number_of_entries = 1
        frame_type = 85 /* same_locals_1_stack_item */
          stack = [ class java/lang/Throwable ]
      LineNumberTable:
        line 3: 4
        line 4: 18
```

sequence of `nop`s followed by `athrow` at offsets `21` - `23` can not be reached.

**Comment by danielsun1106** [comment]
@Godin  It looks that we have more work to complete ;-)  
/cc @paulk-asert

**Comment by paulk-asert** [comment]
Hmm. Looks like we do.

**Comment by admxiii** [comment]
Also, GROOVY-9126 fixed 'return' statements which was a large chunk of the problems, but there are still impossible lines after switch blocks and if/elseif/else blocks and try/catch blocks.

**Comment by danielsun1106** [comment]
![](https://user-images.githubusercontent.com/138671/70068687-c0d8f600-15f0-11ea-8d1b-df394c0a71dd.png)

@Godin The issues you spotted in the above picture will be fixed in Groovy 3.0.7 and 4.0.0 alpha-2. /cc  @paulk-asert 

The related commits are here:
https://github.com/apache/groovy/commit/71ae8e460998c252cde06bfc8d0e059a2f9b7ff9
https://github.com/apache/groovy/commit/6292f7852a8dadb526261b2e54b1017c2aeed5ae
